python_benchmarks/benchmark_softmax.py

```diff
@@ -0,0 +1,160 @@
+    )
+    if dtype is not DataType.Float:
+        T0 = fd.ops.cast(T0, dtype=DataType.Float)
+    T2 = fd.ops.max(T0, axes=[reduction_axis], keepdim=False, dtype=DataType.Null)
+
+    if reduction_axis:
+        V6 = fd.define_vector([T0.size(0), 1], dtype=DataType.Int)
+    else:
+        V6 = fd.define_vector([1, T0.size(1)], dtype=DataType.Int)
+    bcast_dim = 1 - reduction_axis
+
+    T7 = fd.ops.broadcast_in_dim(T2, shape=V6, broadcast_dims=[bcast_dim])
+
+    V11 = T0.shape()
+    T12 = fd.ops.broadcast_in_dim(T7, shape=V11, broadcast_dims=[0, 1])
+    T13 = fd.ops.sub(T0, T12)
+    T14 = fd.ops.exp(T13)
+    T15 = fd.ops.sum(T14, axes=[reduction_axis], keepdim=False, dtype=DataType.Null)
+
+    T20 = fd.ops.broadcast_in_dim(T15, shape=V6, broadcast_dims=[bcast_dim])
+    T25 = fd.ops.broadcast_in_dim(T20, shape=V11, broadcast_dims=[0, 1])
+
+    T26 = fd.ops.reciprocal(T25)
+    T27 = fd.ops.mul(T14, T26)
+
+    if dtype is not DataType.Float:
+        T27 = fd.ops.cast(T27, dtype=dtype)
+    fd.add_output(T27)
+
+
+def softmax_bwd_fusion(
+    fd: FusionDefinition, dtype: DataType, reduction_axis: int
+) -> None:
+    T0 = fd.define_tensor(
+        shape=[-1, -1],
+        contiguity=[True, True],
+        dtype=dtype,
+        is_cpu=False,
+    )
+    T1 = fd.define_tensor(
+        shape=[-1, -1],
+        contiguity=[True, True],
+        dtype=dtype,
+        is_cpu=False,
+    )
+
+    if dtype is not DataType.Float:
+        T0 = fd.ops.cast(T0, dtype=DataType.Float)
+        T1 = fd.ops.cast(T1, dtype=DataType.Float)
+
+    T4 = fd.ops.mul(T0, T1)
+    T5 = fd.ops.sum(T4, axes=[reduction_axis], keepdim=False, dtype=DataType.Null)
+
+    if reduction_axis:
+        V9 = fd.define_vector([T0.size(0), 1], dtype=DataType.Int)
+    else:
+        V9 = fd.define_vector([1, T0.size(1)], dtype=DataType.Int)
+    bcast_dim = 1 - reduction_axis
+
+    T10 = fd.ops.broadcast_in_dim(T5, shape=V9, broadcast_dims=[bcast_dim])
+
+    if dtype is not DataType.Float:
+        T10 = fd.ops.cast(T10, dtype=dtype)
+
+    V15 = fd.define_vector([T0.size(0), T0.size(1)], dtype=DataType.Int)
+    T16 = fd.ops.broadcast_in_dim(T10, shape=V15, broadcast_dims=[0, 1])
+
+    if dtype is not DataType.Float:
+        T16 = fd.ops.cast(T16, dtype=DataType.Float)
+
+    T18 = fd.ops.sub(T1, T16)
+    T19 = fd.ops.mul(T0, T18)
+
+    if dtype is not DataType.Float:
+        T19 = fd.ops.cast(T19, dtype=dtype)
+    fd.add_output(T19)
+
+
+def generate_input_sizes():
+    range_outer = [2**i for i in range(1, 5)]
+    range_inner = [32 * 1024 * 2**i for i in range(11)]
+
+    inputs = [(i, j) for i in range_outer for j in range_inner]
+    inputs.extend([(j, i) for i in range_outer for j in range_inner])
+    return inputs
+
+
+@pytest.mark.parametrize("size", generate_input_sizes())
+@pytest.mark.parametrize("dtype", [torch.float16, torch.float32])
+@pytest.mark.parametrize("reduction_axis", [0, 1])
+def test_softmax_fwd(size, dtype, reduction_axis):
```

Comment:
This was just a correctness test, and can be skipped when running benchmarks alone. I wanted to keep them separate to avoid correctness tests being run everytime the benchmarks are run.

They are useful to verify the fusion definitions. We can add `pytest.disable` to permanently disable them. In the future, if we reiterate over the fusion definition, we may want to verify the correctness of the new definition as well.

Reply:
Yes, it's important to verify results. My question is if we need to define a separate function for verifications. Can we use the benchmark function as well by having it optionally verify the results?

Reply:
True, added a flag to optionally run correctness tests within the benchmark functions.